pkg/controller/kms/key/setup.go

```diff
@@ -0,0 +1,330 @@
+package key
+
+import (
+	"context"
+	"sort"
+
+	svcsdk "github.com/aws/aws-sdk-go/service/kms"
+	svcsdkapi "github.com/aws/aws-sdk-go/service/kms/kmsiface"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/kms/v1alpha1"
+	aws "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+// SetupKey adds a controller that reconciles Key.
+func SetupKey(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(svcapitypes.KeyGroupKind)
+	opts := []option{
+		func(e *external) {
+			e.preObserve = preObserve
+			e.postObserve = postObserve
+			e.preCreate = preCreate
+			e.postCreate = postCreate
+			u := &updater{client: e.client}
+			e.update = u.update
+			d := &deleter{client: e.client}
+			e.delete = d.delete
+			o := &observer{client: e.client}
+			e.isUpToDate = o.isUpToDate
+			e.lateInitialize = o.lateInitialize
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&svcapitypes.Key{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.KeyGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+func preObserve(_ context.Context, cr *svcapitypes.Key, obj *svcsdk.DescribeKeyInput) error {
+	obj.KeyId = aws.String(meta.GetExternalName(cr))
+	return nil
+}
+
+func postObserve(_ context.Context, cr *svcapitypes.Key, obj *svcsdk.DescribeKeyOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return obs, err
+	}
+
+	// Set Condition
+	switch aws.StringValue(obj.KeyMetadata.KeyState) {
+	case string(svcapitypes.KeyState_Enabled):
+		cr.SetConditions(xpv1.Available())
+	case string(svcapitypes.KeyState_Disabled):
+		cr.SetConditions(xpv1.Unavailable())
+	case string(svcapitypes.KeyState_PendingDeletion):
+		cr.SetConditions(xpv1.Deleting())
+	case string(svcapitypes.KeyState_PendingImport):
+		cr.SetConditions(xpv1.Unavailable())
+	case string(svcapitypes.KeyState_Unavailable):
+		cr.SetConditions(xpv1.Unavailable())
+	}
+
+	// special case: if key is scheduled for deletion, abort early and do not schedule for deletion again
+	if obj.KeyMetadata.DeletionDate != nil {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	return obs, nil
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.Key, obj *svcsdk.CreateKeyInput) error {
+	return nil
+}
```

Comment:
```suggestion
```

You don't have to supply all hooks. They all have no-op defaults in `controller.go`.